### PR TITLE
[SPARK-15982][SPARK-16009][SPARK-16007][SQL] Harmonize the behavior of DataFrameReader.text/csv/json/parquet/orc

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?

Issues with current reader behavior.
- `text()` without args returns an empty DF with no columns -> inconsistent, its expected that text will always return a DF with `value` string field,
- `textFile()` without args fails with exception because of the above reason, it expected the DF returned by `text()` to have a `value` field.
- `orc()` does not have var args, inconsistent with others
- `json(single-arg)` was removed, but that caused source compatibility issues - [SPARK-16009](https://issues.apache.org/jira/browse/SPARK-16009)
- user specified schema was not respected when `text/csv/...` were used with no args - [SPARK-16007](https://issues.apache.org/jira/browse/SPARK-16007)

The solution I am implementing is to do the following. 
- For each format, there will be a single argument method, and a vararg method. For json, parquet, csv, text, this means adding json(string), etc.. For orc, this means adding orc(varargs).
- Remove the special handling of text(), csv(), etc. that returns empty dataframe with no fields. Rather pass on the empty sequence of paths to the datasource, and let each datasource handle it right. For e.g, text data source, should return empty DF with schema (value: string)
- Deduped docs and fixed their formatting.
## How was this patch tested?

Added new unit tests for Scala and Java tests

### CODE REVIEW

**Comment by tdas** [comment]
@marmbrus @rxin

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 1]
```diff
@@ -119,13 +119,7 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
    * @since 1.4.0
    */
   def load(): DataFrame = {
-    val dataSource =
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: deduped.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 2]
```diff
@@ -146,18 +140,15 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
    */
   @scala.annotation.varargs
   def load(paths: String*): DataFrame = {
-    if (paths.isEmpty) {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: removed the special handling of empty paths

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 3]
```diff
@@ -276,7 +267,42 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
    * `spark.sql.columnNameOfCorruptRecord`): allows renaming the new field having malformed string
    * created by `PERMISSIVE` mode. This overrides `spark.sql.columnNameOfCorruptRecord`.</li>
    *
-   * @since 1.6.0
+   * @since 1.4.0
+   */
+  def json(path: String): DataFrame = json(Seq(path): _*)
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: made this method depend on `json(varargs)` to prevent code duplication.

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: we should add some comment inline on why this exists. ditto for similar functions.

Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: yes. agreed. will add as inline comments, not as scala docs.

**Comment by gatorsmile** [comment]
The following test case is for SPARK-16007. I can confirm this PR already fixes the issue.
```scala
  test("schema checking") {
    val schema: StructType = new StructType().add("s", "string")
    assert(spark.read.schema(schema).csv().schema === schema)
    assert(spark.read.schema(schema).json().schema === schema)
    assert(spark.read.schema(schema).parquet().schema === schema)
    assert(spark.read.schema(schema).text().schema === schema)
    assert(spark.read.schema(schema).orc().schema === schema)
  }
```

Since the ORC data source must be used with Hive support enabled, you can comment the last line out. Or move it to another test case in Hive. 

Please let me know if anything is needed. Thanks!

**Comment by tdas** [comment]
@gatorsmile I updated my tests to rigorously test this. However since this PR is about testing the common behavior (e.g. whether all of them respect user schema or not), I have not added tests for the case where there are no paths AND no user schema. The behavior will be source-specific - `parquet/json/csv` should throw error, whereas `text` should not. Could you make a PR that tests these in  CSVSuite, etc. if they are not already tested?

**Comment by gatorsmile** [comment]
@tdas Sure, will do it soon. I might submit it after this is merged. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 4]
```diff
@@ -368,6 +397,63 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
...
+   * value.</li>
+   * <li>`dateFormat` (default `null`): sets the string that indicates a date format. Custom date
+   * formats follow the formats at `java.text.SimpleDateFormat`. This applies to both date type
+   * and timestamp type. By default, it is `null` which means trying to parse times and date by
+   * `java.sql.Timestamp.valueOf()` and `java.sql.Date.valueOf()`.</li>
+   * <li>`maxColumns` (default `20480`): defines a hard limit of how many columns
+   * a record can have.</li>
+   * <li>`maxCharsPerColumn` (default `1000000`): defines the maximum number of characters allowed
+   * for any given value being read.</li>
+   * <li>`mode` (default `PERMISSIVE`): allows a mode for dealing with corrupt records
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: This can be addressed in a follow up, but I don't think we should duplicate docs cause they are going to get out of sync.  I'd have one canonical one and the other link to it.

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaDataFrameReaderWriterSuite.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    spark.read().text();
+    spark.read().text(input);
+    spark.read().text(input, input, input);
+    spark.read().text(new String[]{input, input})
+        .write().text(output);
+  }
+
+  @Test
+  public void testTextFileAPI() {
+    spark.read().textFile();     // Disabled because of SPARK-XXXXX
```
Comment (marmbrus) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDataFrameReaderWriterSuite.java]: SPARK-XXXX?

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 6]
```diff
@@ -228,4 +220,101 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext {
...
+    Option(input).map(spark.read.format("org.apache.spark.sql.test").load)
+  }
+
+  test("text - API and common behavior") {
+    // Reader, without user specified schema
+    assert(spark.read.text().schema === textSchema)
+    assert(spark.read.text(input).schema === textSchema)
+    assert(spark.read.text(input, input, input).schema === textSchema)
+    assert(spark.read.text(Seq(input, input): _*).schema === textSchema)
+    assert(Option(input).map(spark.read.text).get.schema === textSchema) // SPARK-16009
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: Nit: its great to include a super short description with the JIRA.

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 7]
```diff
@@ -228,4 +220,101 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext {
...
+
+    // Reader, with user specified schema
+    assert(spark.read.schema(userSchema).text().schema === userSchema)
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: I'm not sure what it means to read text with a user specified schema.  In fact, this can't actually work if there was data and you call `collect()`.

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 8]
```diff
@@ -228,4 +220,101 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext {
...
+    // Test explicit calls to single arg method - SPARK-16009
+    assert(Option(input).map(spark.read.schema(userSchema).parquet).get.schema === userSchema)
+
+    // Writer
+    spark.range(10).write.parquet(output)
+  }
+
+  /**
+   * This only tests whether API compiles, but does not run it as orc()
+   * cannot be run with Hive classes.
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: nit: `without`

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 9]
```diff
@@ -228,4 +220,101 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext {
...
+    assert(spark.read.schema(userSchema).parquet().schema === userSchema)
+    assert(spark.read.schema(userSchema).parquet(input).schema === userSchema)
+    assert(spark.read.schema(userSchema).parquet(input, input, input).schema === userSchema)
+    assert(spark.read.schema(userSchema).parquet(Seq(input, input): _*).schema === userSchema)
+
+    // Test explicit calls to single arg method - SPARK-16009
+    assert(Option(input).map(spark.read.schema(userSchema).parquet).get.schema === userSchema)
+
+    // Writer
+    spark.range(10).write.parquet(output)
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: should we checkAnswer on these there possible?  a lot of errors only manifest on `collect()`

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 10]
```diff
@@ -276,7 +267,45 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
...
+   * <li>`allowSingleQuotes` (default `true`): allows single quotes in addition to double quotes
+   * </li>
+   * <li>`allowNumericLeadingZeros` (default `false`): allows leading zeros in numbers
+   * (e.g. 00012)</li>
+   * <li>`allowBackslashEscapingAnyCharacter` (default `false`): allows accepting quoting of all
+   * character using backslash quoting mechanism</li>
+   * <li>`mode` (default `PERMISSIVE`): allows a mode for dealing with corrupt records
+   * during parsing.</li>
+   * <ul>
+   *  <li>`PERMISSIVE` : sets other fields to `null` when it meets a corrupted record, and puts the
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: this does not indent correctly:

screen shot 2016-06-17 at 2 07 18 pm

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: It might be better to just use the intented `-` notation for lists

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 11]
```diff
@@ -228,4 +220,101 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext {
...
+    // Writer
+    spark.read.text().write.text(output)
+  }
+
+  test("textFile - API and common behavior") {
+    // Reader, without user specified schema
+    assert(spark.read.textFile().schema === textSchema)
+    assert(spark.read.textFile(input).schema === textSchema)
+    assert(spark.read.textFile(input, input, input).schema === textSchema)
+    assert(spark.read.textFile(Seq(input, input): _*).schema === textSchema)
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: What happens if you specify a schema here?

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala (hunk 9 above)
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: My goal in these tests were to test the API, and individual suites like CsvSuite, ParquetSuite, etc should take care of correctness. But I see your point, I could add a few basic tests.

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala (hunk 7 above)
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: In that case, then its probably better for stuff to fail earlier, that is TextFileFormat should fail if there is a userSpecifiedSchema.

**Comment by marmbrus** [comment]
A few comments. Overall LGTM.

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala (hunk 9 above)
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: Yeah, only where its easy.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: In my PR, will add the test cases to verify all the possible inputs after this code changes. Thanks!

**Comment by tdas** [comment]
@marmbrus I deduped the docs. Linking to specific method using scaladocs was hard to use and did not work in Java docs. So I just wrote it "See docs on other overloaded method". And I fixed the doc formatting.

**Comment by zsxwing** [comment]
LGTM. Merging to master and 2.0. Thanks!

**Comment by gatorsmile** [comment]
Tonight, will submit a PR for test cases. Thanks!

**Comment by tdas** [comment]
@gatorsmile Actually, I already added tests in this PR that cover the scenario where schema is not provided. The only one that is not really tested in the orc, as it cannot be run in the DataFrameReaderWriterSuite. So could you add that test in the ORC-related test suites, if at all needed?

**Comment by gatorsmile** [comment]
@tdas Sure, I have multiple related test cases I wrote a few days ago. It might be also useful. You can judge whether they are needed or not. Let me merge your latest changes. : )

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 12]
```diff
@@ -228,4 +222,152 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext {
...
+    testRead(spark.read.parquet(dir, dir), data ++ data, schema)
+    testRead(spark.read.parquet(Seq(dir, dir): _*), data ++ data, schema)
+    // Test explicit calls to single arg method - SPARK-16009
+    testRead(Option(dir).map(spark.read.parquet).get, data, schema)
+
+    // Reader, with user specified schema, data should be nulls as schema in file different
+    // from user schema
+    val expData = Seq[String](null, null, null)
+    testRead(spark.read.schema(userSchema).parquet(), Seq.empty, userSchema)
+    testRead(spark.read.schema(userSchema).parquet(dir), expData, userSchema)
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: @tdas ORC behaves differently. When the user-specified schema does not match the physical schema, it simply stops and reports an exception. Do you think that behavior is better than returning `null` for all the rows?

Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: Let's documents that as a test for now.

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: Sure, will do it. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 13]
```diff
@@ -135,7 +129,7 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
    * @since 1.4.0
    */
   def load(path: String): DataFrame = {
-    option("path", path).load()
+    load(Seq(path): _*) // force invocation of `load(...varargs...)`
```
Comment (koertkuipers) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: with this change path is no longer available in the options. this makes it hard (impossible?) for non-file based DataSources (not implementing FileFormat) to use load(...)

For example for elasticsearch we use:
```
sqlContext.read.format("org.elasticsearch.spark.sql").load(resource)
```
i do not think this can be implemented anymore now?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]:
```Scala
sqlContext.read.option("path", resource).format("org.elasticsearch.spark.sql").load()
```

Can you try this?

Comment (koertkuipers) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: i believe that works as expected (i am running into some other issues now, but they seem unrelated). 
however from a DSL perspective this is not very pretty?

Comment (koertkuipers) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: it will also break users code in an upgrade

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: yea this is a bad breaking change.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: Do you want me to fix it?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: If you can!

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: Sure, will do it soon. Thanks!
